### DISCUSSION TITLE
Ideas for the future: Export application

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#31] [Ideas]
I was thinking about @tabreturn's comments about portability, like packaging a self contained modified Thonny IDE, and also thinking about this keynote by Russell Keith-Magee (of beeware.org fame) I re-watched recently https://youtu.be/ftP5BQh1-YM?t=1253

Sending your project to non-programmer friends is a big challenge. 
If we could have some helpers/templates for that it would be great. I see some paths one could explore:

- Pyinstaller https://stackoverflow.com/questions/24049391/how-to-bundle-jar-files-with-pyinstaller
- http://Beeware.org (very experimental)
- Jupyter Notebooks based sharing tools, I think someone mentioned something about it here, mybinder.org can handle IJava https://github.com/SpencerPark/IJava I'm not sure how to have both Java and Python simultaneously.

**Comment by hx2A** [comment]
All great ideas. I agree, sending Python projects in particular to non-programmers can be a big challenge.

I have used Pyinstaller before and have coded some of the classpath stuff to accommodate Pyinstaller's needs, but I haven't gotten around to giving it a try. Theoretically this should be possible but I know it would require recipients to have Java 11 installed already. Unless if that could be packaged along with the other files...I do fear that if this worked the package size would be huge.

I already have py5 working on mybinder:

https://mybinder.org/v2/gh/hx2A/py5examples/HEAD?urlpath=lab

Don't try py5bot through there because I inexplicably botched that part of the release. The other stuff works though. You might be interested in the Py5SketchPortal.

**Reply by hx2A** [reply]
Actually, py5bot does work through there. I just tested it. But it might not work if you try to run it locally.

**Reply by tabreturn** [reply]
> Theoretically this should be possible but I know it would require recipients to have Java 11 installed already. Unless if that could be packaged along with the other files...I do fear that if this worked the package size would be huge.

Processing's PDE is bundled with a Java runtime and [it's ~130 MB](https://www.processing.org/download/). 

Thonny's portable version is [~30 MB](https://github.com/thonny/thonny/releases/); my py5 virtual env zips into ~90 MB; OpenJDK is ~180 MB. So that's ~300 MB, but I'm sure there ways to reduce this (I suspect one wouldn't need Jupyter for a Thonny bundle, and you don't need all of OpenJDK, etc.)

Thanks for the mybinder link! :champagne: 

EDIT: a [py5 plug-in for Thonny](https://github.com/tabreturn/thonny-py5mode) will break up the install process into an initial Thonny download (~30 MB), then the rest is downloaded upon activating py5 through the Thonny plug-in manager.

**Reply by hx2A** [reply]
> I suspect one wouldn't need Jupyter for a Thonny bundle

True, but one of py5's package requirements is the ipykernel package, and that requires a whole web of other packages. I'm not sure if Jupyter is in there.
